File: languages/php/mandelbrot-par.php
<?php

// mandelbrot-par.php: parallel mandelbrot via pcntl_fork, row-band decomposition.
// Invocation: php mandelbrot-par.php <cores> <n>
// Output: identical to mandelbrot.php at the same n (core-invariant).
//
// Each worker owns rows [row_start, row_end) and sends back a single int64 (the
// in-set pixel count for its band) via a pipe. The parent sums all band counts.

function worker_count(int $y_start, int $y_end, int $n): int {
    $count = 0;
    for ($y = $y_start; $y < $y_end; $y++) {
        $ci = 2.0 * $y / $n - 1.0;
        for ($x = 0; $x < $n; $x++) {
            $cr = 2.0 * $x / $n - 1.5;
            $zr = 0.0;
            $zi = 0.0;
            $tr = 0.0;
            $ti = 0.0;
            $i = 0;
            while ($i < 50 && $tr + $ti <= 4.0) {
                $t  = $zr * $zi;
                $zi = $t + $t + $ci;   // 2*zr*zi + ci, FMA-proof
                $zr = $tr - $ti + $cr;
                $tr = $zr * $zr;
                $ti = $zi * $zi;
                $i++;
            }
            if ($tr + $ti <= 4.0) $count++;
        }
    }
    return $count;
}

$cores = isset($argv[1]) ? (int)$argv[1] : 1;
$n     = isset($argv[2]) ? (int)$argv[2] : 128;

if ($cores <= 1) {
    // Serial fallback: identical to mandelbrot.php
    $t0 = hrtime(true);
    $count = worker_count(0, $n, $n);
    $ns = hrtime(true) - $t0;
    fwrite(STDERR, "COMPUTE_NS $ns\n");
    echo $count, "\n";
    echo "mandelbrot($n)\n";
    exit(0);
}

// Fork $cores workers; each writes an 8-byte packed int64 (LE) to its pipe.
$pipes = [];
$pids  = [];

$t0 = hrtime(true);
for ($w = 0; $w < $cores; $w++) {
    $row_start = intdiv($w * $n, $cores);
    $row_end   = intdiv(($w + 1) * $n, $cores);

    // pipe: [0]=read end (parent), [1]=write end (child)
    $pair = stream_socket_pair(STREAM_PF_UNIX, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
    if ($pair === false) {
        fwrite(STDERR, "stream_socket_pair failed\n");
        exit(1);
    }

    $pid = pcntl_fork();
    if ($pid < 0) {
        fwrite(STDERR, "pcntl_fork failed\n");
        exit(1);
    }

    if ($pid === 0) {
        // Child: close read end, compute, write result, exit.
        fclose($pair[0]);
        $count = worker_count($row_start, $row_end, $n);
        fwrite($pair[1], pack('P', $count));  // 8 bytes, unsigned 64-bit LE
        fclose($pair[1]);
        exit(0);
    }

    // Parent: close write end, keep read end.
    fclose($pair[1]);
    $pipes[$w] = $pair[0];
    $pids[$w]  = $pid;
}

// Parent: collect results in worker order (guaranteed order = core-invariant).
$total = 0;
for ($w = 0; $w < $cores; $w++) {
    $raw = stream_get_contents($pipes[$w]);
    fclose($pipes[$w]);
    $v = unpack('P', $raw);  // 'P' = machine-endian unsigned 64-bit
    $total += $v[1];
}

// Reap children.
for ($w = 0; $w < $cores; $w++) {
    pcntl_waitpid($pids[$w], $status);
}
$ns = hrtime(true) - $t0;
fwrite(STDERR, "COMPUTE_NS $ns\n");

echo $total, "\n";
echo "mandelbrot($n)\n";

File: languages/perl/mandelbrot-par.pl
use strict;
use warnings;
use Time::HiRes qw(clock_gettime CLOCK_MONOTONIC);

# mandelbrot-par: parallel variant of the mandelbrot benchmark.
# Invocation: perl mandelbrot-par.pl <cores> <n>
# Decomposes the N rows into `cores` horizontal bands; each worker (fork)
# computes the in-set count for its band and returns it to the parent via a
# pipe.  Output is bit-identical to the serial benchmark for any core count.

sub mandelbrot_band {
    my ($n, $row_start, $row_end) = @_;
    my $count = 0;
    for my $y ($row_start .. $row_end - 1) {
        my $ci = 2.0 * $y / $n - 1.0;
        for my $x (0 .. $n - 1) {
            my $cr = 2.0 * $x / $n - 1.5;
            my ($zr, $zi, $tr, $ti) = (0.0, 0.0, 0.0, 0.0);
            my $i = 0;
            while ($i < 50 && $tr + $ti <= 4.0) {
                my $t = $zr * $zi;
                $zi = $t + $t + $ci;   # == 2*zr*zi + ci, FMA-proof
                $zr = $tr - $ti + $cr;
                $tr = $zr * $zr;
                $ti = $zi * $zi;
                $i++;
            }
            $count++ if $tr + $ti <= 4.0;
        }
    }
    return $count;
}

my $cores = @ARGV >= 1 ? int($ARGV[0]) : 1;
my $n     = @ARGV >= 2 ? int($ARGV[1]) : 128;

if ($cores == 1) {
    # Inline path: no fork overhead.
    my $t0 = clock_gettime(CLOCK_MONOTONIC);
    my $count = mandelbrot_band($n, 0, $n);
    my $ns = int((clock_gettime(CLOCK_MONOTONIC) - $t0) * 1e9);
    print STDERR "COMPUTE_NS $ns\n";
    print "$count\n";
    print "mandelbrot($n)\n";
    exit 0;
}

# Fork one child per band.  Each child writes its count as a decimal line to
# its write-end of a pipe and exits.  Parent reads from all pipes in order,
# sums, then prints.
my @pipes;   # $pipes[$w] = read filehandle for worker $w

my $t0 = clock_gettime(CLOCK_MONOTONIC);

for my $w (0 .. $cores - 1) {
    my $row_start = int($w * $n / $cores);
    my $row_end   = int(($w + 1) * $n / $cores);

    pipe(my $rh, my $wh) or die "pipe: $!";

    my $pid = fork();
    die "fork: $!" unless defined $pid;

    if ($pid == 0) {
        # Child
        close $rh;
        my $count = mandelbrot_band($n, $row_start, $row_end);
        print $wh "$count\n";
        close $wh;
        exit 0;
    }

    # Parent
    close $wh;
    $pipes[$w] = $rh;
}

# Collect results in band order (preserves deterministic ordering).
my $total = 0;
for my $w (0 .. $cores - 1) {
    my $line = readline($pipes[$w]);
    chomp $line;
    $total += int($line);
    close $pipes[$w];
}

# Wait for all children.
for (1 .. $cores) { wait() }

my $ns = int((clock_gettime(CLOCK_MONOTONIC) - $t0) * 1e9);
print STDERR "COMPUTE_NS $ns\n";

print "$total\n";
print "mandelbrot($n)\n";
